File: min_stack.js
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

var MinStack = function () {
  this.stack = [];
  this.min = Infinity;
  this.stack.push(this.min);
};

/**
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function (val) {
  if (val <= this.min) {
    this.stack.push(this.min);
    this.min = val;
  }
  this.stack.push(val);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
  let poped = this.stack.pop();
  if (poped == this.min) {
    this.min = this.stack.pop();
  }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  return this.stack[this.stack.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  return this.min;
};

test("Scenario #1:", () => {
  var obj = new MinStack();
  obj.push(5);
  obj.push(6);
  obj.push(7);
  obj.pop();
  expect(obj.top()).toStrictEqual(6);
  expect(obj.getMin()).toStrictEqual(5);
});

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: Min Stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minstack = []
        self.min = int(sys.maxsize)
        self.minstack.append(self.min)

    # def push(self, val: int) -> None:
    #     if val <= self.min:
    #         self.min = val
        
    #     self.stack.append(val)   
    #     self.minstack.append(self.min) 

    def push(self, val: int) -> None:
        if val <= self.min:
            self.stack.append(self.min) 
            self.min = val
            
        self.stack.append(val)   
        

    # def pop(self) -> None:
    #     self.stack.pop()
    #     self.minstack.pop()
    #     self.min = self.minstack[-1] 

    def pop(self) -> None:
        popped = self.stack.pop()
        if popped == self.min:
            self.min = self.stack.pop()

        
          

    def top(self) -> int:
        return self.stack[-1] 
        

    def getMin(self) -> int:
        return self.min
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()